### PR TITLE
Add `blacken-docs` and `pyupgrade` to pre-commit configuration

### PR DESCRIPTION
**Pull request by Takishima**

### CODE REVIEW

**Comment by Takishima** [comment]
NB: need to wait for #440 to be merged to address the CI failures.

File: projectq/setups/decompositions/amplitudeamplification_test.py [hunk 1]
```diff
@@ -90,7 +89,7 @@ def test_simple_grover():
 
     assert total_prob_after == pytest.approx(
         theoretical_prob, abs=1e-6
-    ), "The obtained probability is less than expected %f vs. %f" % (
+    ), "The obtained probability is less than expected {:f} vs. {:f}".format(
```
Comment (andreashehn) [inline projectq/setups/decompositions/amplitudeamplification_test.py:92]: Why not use the `f'{total_prob_after:f}` formatting here?

File: projectq/setups/decompositions/amplitudeamplification_test.py [hunk 2]
```diff
@@ -172,7 +171,7 @@ def test_complex_aa():
 
     assert total_prob_after == pytest.approx(
         theoretical_prob, abs=1e-2
-    ), "The obtained probability is less than expected %f vs. %f" % (
+    ), "The obtained probability is less than expected {:f} vs. {:f}".format(
```
Comment (andreashehn) [inline projectq/setups/decompositions/amplitudeamplification_test.py:174]: Why not use the `f'{total_prob_after:f}` formatting here?

File: projectq/setups/decompositions/phaseestimation_test.py [hunk 3]
```diff
@@ -58,7 +57,7 @@ def test_simple_test_X_eigenvectors():
         eng.flush()
 
     num_phase = (results == 0.5).sum()
-    assert num_phase / N >= 0.35, "Statistics phase calculation are not correct (%f vs. %f)" % (
+    assert num_phase / N >= 0.35, "Statistics phase calculation are not correct ({:f} vs. {:f})".format(
```
Comment (andreashehn) [inline projectq/setups/decompositions/phaseestimation_test.py:60]: Why not use the `f'{total_prob_after:f}` formatting here?

File: projectq/setups/decompositions/phaseestimation_test.py [hunk 4]
```diff
@@ -91,7 +90,7 @@ def test_Ph_eigenvectors():
         eng.flush()
 
     num_phase = (results == 0.125).sum()
-    assert num_phase / N >= 0.35, "Statistics phase calculation are not correct (%f vs. %f)" % (
+    assert num_phase / N >= 0.35, "Statistics phase calculation are not correct ({:f} vs. {:f})".format(
```
Comment (andreashehn) [inline projectq/setups/decompositions/phaseestimation_test.py:93]: Why not use the `f'{total_prob_after:f}` formatting here?

File: projectq/setups/decompositions/phaseestimation_test.py [hunk 5]
```diff
@@ -129,7 +128,7 @@ def test_2qubitsPh_andfunction_eigenvectors():
         eng.flush()
 
     num_phase = (results == 0.125).sum()
-    assert num_phase / N >= 0.34, "Statistics phase calculation are not correct (%f vs. %f)" % (
+    assert num_phase / N >= 0.34, "Statistics phase calculation are not correct ({:f} vs. {:f})".format(
```
Comment (andreashehn) [inline projectq/setups/decompositions/phaseestimation_test.py:131]: Why not use the `f'{total_prob_after:f}` formatting here?

File: projectq/setups/decompositions/phaseestimation_test.py [hunk 6]
```diff
@@ -179,7 +178,7 @@ def test_X_no_eigenvectors():
     assert total == pytest.approx(N, abs=5)
     assert plus_probability == pytest.approx(
         1.0 / 4.0, abs=1e-1
-    ), "Statistics on |+> probability are not correct (%f vs. %f)" % (
+    ), "Statistics on |+> probability are not correct ({:f} vs. {:f})".format(
```
Comment (andreashehn) [inline projectq/setups/decompositions/phaseestimation_test.py:181]: Why not use the `f'{total_prob_after:f}` formatting here?

File: setup.py [hunk 7]
```diff
@@ -139,7 +138,7 @@ def compiler_test(
         if compiler.compiler_type == 'msvc':
             err.close()
             os.dup2(olderr, sys.stderr.fileno())
-            with open('err.txt', 'r') as err_file:
+            with open('err.txt') as err_file:
```
Comment (andreashehn) [inline setup.py:141]: Why was the `'r'` removed here?

**Review by andreashehn — APPROVED** [review]
LGTM overall. Some minor comments.

File: docs/conf.py [hunk 8]
```diff
@@ -560,10 +560,10 @@ def linkcode_resolve(domain, info):
     os.mkdir(docgen_path)
 
 for desc in descriptions:
-    fname = os.path.join(docgen_path, 'projectq.{}.rst'.format(desc.name))
+    fname = os.path.join(docgen_path, f'projectq.{desc.name}.rst')
     lines = None
     if os.path.exists(fname):
-        with open(fname, 'r') as fd:
+        with open(fname) as fd:
```
Comment (andreashehn) [inline docs/conf.py:566]: Why remove the `'r'`?

Comment (Takishima) [reply docs/conf.py:566]: According to the Python standard, `'r'` is the default value for the mode. And therefore removed by `pyupgrade`

File: setup.py (hunk 7 above)
Comment (Takishima) [reply setup.py:141]: According to the Python standard, `'r'` is the default value for the mode. And therefore removed by `pyupgrade`

File: docs/conf.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Takishima) [reply docs/conf.py:566]: According to the Python standard, `'r'` is the default value for the mode. And therefore removed by `pyupgrade`

File: projectq/setups/decompositions/phaseestimation_test.py (hunk 6 above)
Comment (Takishima) [reply projectq/setups/decompositions/phaseestimation_test.py:181]: Will be addressed in the next commit.

File: projectq/setups/decompositions/phaseestimation_test.py (hunk 5 above)
Comment (Takishima) [reply projectq/setups/decompositions/phaseestimation_test.py]: Will be addressed in the next commit.

File: projectq/setups/decompositions/phaseestimation_test.py (hunk 4 above)
Comment (Takishima) [reply projectq/setups/decompositions/phaseestimation_test.py]: Will be addressed in the next commit.

**Pull request merged**